---
title: "Transcriptomics and RNA-Seq"
author: "Jason Patrick Bennett"
date: "May 17, 2018"
output: 
  html_document:
    keep_md: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Transcriptomics: RNA-Seq Continued

First lets grab our file:

```{r}
counts <- read.csv("data/airway_scaledcounts.csv", stringsAsFactors = FALSE)
```

And now lets grab our metadata:

```{r}
metadata <- read.csv("data/airway_metadata.csv", stringsAsFactors = FALSE)
```

To learn how to do basic differential expression analysis, we will use this sample data. **This is not the way we would do differential expression analysis normally!**

Let's find the mean amount of counts between the genes:

```{r}
# This grabs only the control samples from our Metadata
control <- metadata[metadata[,"dex"]=="control",]

# Calculate the mean
control.mean <- rowSums( counts[,control$id] ) / nrow(control)

# This will name the samples based on the names from the original Counts data
names(control.mean) <- counts$ensgene
```

*Question 1: How would we make the code more robust?*
  -We could replace the "4" with the count of the rows of "control"(example seen above)
  
*Question 2: Follow the same procedure for the treated samples:*
  -See following code
  
```{r}
treated <- metadata[metadata[,"dex"]=="treated",]
treated.mean <- rowSums( counts[,treated$id] ) / nrow(treated)
names(treated.mean) <- counts$ensgene
```

Now lets combine our meancount data just to collate our data:

```{r}
meancounts <- data.frame(control.mean, treated.mean)
```

Now lets inspect our new meancounts dataframe:

```{r}
colSums(meancounts)
```

And lets use a scatterplot to help us visualize the data:

```{r}
# We used the log option here to show all the data we need
plot(meancounts$control.mean, meancounts$treated.mean, log = "xy", xlab = "Control", ylab = "Treated")
```

Now lets look for genes with a large change between control and treated samples. This is generally done by loooking at the log2 of the fold change (for mathematical reasons).

Lets calculate the log2foldchange, then add that to our meancouts dataframe:

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])

# Inspect meancounts to see if we added the column
head(meancounts)
```

We notice that the data contains positives and negatives. These represent the affect the treatment or non-treatment has had on expression. The *positive* results represent an increase in expression, while the *negative* results represent a decrease in expression.

We can see that there are some strange results:

  -NaN: Results when you divide by zero and attempt to take the log
  -InF: Trying to take the log of zero
  
We need to remove the data from those genes that showed zero expression! We need to filter our data:

```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind = TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

*Question 4: What is the purpose of the arr.ind arguement in the which() function call?*
  -This indicates whether or not the array indices should be returned when the           meancounts[,1:2] == 0 is found within the array
  
We can use a threshold for determining if something actually is differentially expressed by looking at the log2(FoldChange). If the reults is greater than 2, or less than -2, then we probably have conclusive data.

Let's see which genes are up-regulated and which are down-regulated:

```{r}
up.ind <- mycounts$log2fc > 2
upCount <- sum(up.ind)
down.ind <- mycounts$log2fc < (-2)
downCount <- sum(down.ind)

# And a nice way of displaying the data:
paste("The number of UP-regulated genes:", upCount)
paste("The number of DOWN-regulated genes:", downCount)
```

*Question 5: Using up.ind and down.ind, how many up-regulated and down-regulated genes do we have?*
  -Up: 250
  -Down: 367
  
This represents 617 total differentially expressed genes (in either direction). Quite a lot!

## Adding Annotation Data

Lets add annotation to our data now:

```{r}
anno <- read.csv("data/annotables_grch38.csv", stringsAsFactors = FALSE)

# Check the contents
head(anno)
```

We want this annotated data mapped (or **merged**) with the mycounts data. Maybe we should use the merge() function? (Yes. Yes we should.)

*Question 6: Can we set the by.x and by.y arguments in the merge() function to appropriately annotate the mycounts dataframe?*
  -See code below
  
```{r}
results <- merge(mycounts, anno, by.x = "row.names", by.y = "ensgene")

results
```

Now we have our annotated data! Usually though, this is done through Bioconductor:

```{r}
#source("http://bioconductor.org/biocLite.R")
#biocLite()

#biocLite("org.Hs.eg.db")

library("AnnotationDbi")
library("org.Hs.eg.db")
```

```{r}
columns(org.Hs.eg.db)
```

# DESeq2 Analysis

Load the DESeq2 package

```{r}
library(DESeq2)
citation("DESeq2")
```

Now lets prepate our DESeq Data Set Object:

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = metadata,
                              design = ~dex,
                              tidy = TRUE)

dds
```

Lets run the DESeq pipeline on our dataset:

```{r}
sizeFactors(dds)
dispersions(dds)
dds <- DESeq(dds)

res <- results(dds)
res
```

The "padj" column is the p-value that has been adjusted by the DESeq algorithm to account for false positives. Compare this to the "pvalue" to see what our data is actually telling us.

Now lets look at some basic tallies of the data from our results:

```{r}
summary(res)
```

Now lets order our data by the sallest pvalue:

```{r}
# The 'order' command here returns the indices of the sorted data, but doesn't actually sort them. This is important because it allows us to reorder our entire dataframe and not just one part of our data!
resOrdered <- res[order(res$pvalue),]
```

The default pvalue cutoff is 0.1, however we know that in general, we want a p-value around 0.05. We can account for this by applying our own filter:

```{r}
res05 <- results(dds, alpha = 0.05)
summary(res05)
```

*Question 9: How many significant results did we find with a p-value of < 0.05? < 0.01? Save the later as a set of results called resSig01.*
  -See below code
  
```{r}
res01 <- results(dds, alpha = 0.01)
summary(res01)
```

*Question 10: Add annotations to our res01 dataframe.*
  -See below code
  
```{r}
res01.df <- as.data.frame(res01)
resSig01 <- merge(res01.df, anno, by.x = "row.names", by.y = "ensgene")
```

Now lets view our results by the adjusted p-value:

```{r}
ord <- order(resSig01$padj)
head(resSig01[ord,])
```

Finally lets write out our significant results with our added annotations:

```{r}
write.csv(resSig01[ord,], "signif01_results.csv")
```


# Data Visualization

The DESeq2 package offers a function that takes the DESeqDataSet object that we created, the name of a gene, and the name of a variable in the colData that we're interested in.

Lets inspect the gene ID for the CRISPLD2 gene:

```{r}
i <- grep("CRISPLD2", resSig01$symbol)
resSig01[i,]

rownames(resSig01[i,])
```

